src/common/adapters/notion.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * notion.so adapter
+ *
+ * The adapter extracts the UUID of a selected notion.so task ("page") from the
+ * page URL and uses the notion.so API to retrieve the corresponding task
+ * information.
+ *
+ * Note: this adapter uses notion.so's internal v3 API as there is no public
+ * API yet.
+ */
+
+import { match } from 'micro-match';
+
+import client from '../client';
+
+/**
+ * Turns a slugged page ID without dashes into a dasherized RFC 4122 UUID.
+ * UUID-Format: 96ec637d-e4b0-4a5e-acf3-8d4d9a1b2e4b
+ */
+function uuid(slugId) {
+  return [
+    slugId.substring(0, 8),
+    slugId.substring(8, 12),
+    slugId.substring(12, 16),
+    slugId.substring(16, 20),
+    slugId.substring(20),
+  ].join('-');
+}
+
+function getPageFromPath(path) {
+  const { slug } = match('/:organization/:slug', path);
+  if (!slug) return null;
+
+  return slug.replace(/.*-/, ''); // strip title from slug
+}
+
+function getSelectedPageId(loc) {
+  const { pathname: path, searchParams: params } = new URL(loc.href);
+  const isPageModal = params.has('p');
+  const slugId = isPageModal ? params.get('p') : getPageFromPath(path);
+
+  if (!slugId) return null;
+  return uuid(slugId);
+}
+
+function extractTicketInfo(result) {
+  const { value } = result;
+  if (!value) return null;
+
+  const { id, type } = value;
+  if (type !== 'page') return null;
+
+  const title = value.properties.title[0][0];
+  return { id, title, type };
+}
+
+function getTickets(response) {
+  const { results = [] } = response;
+  return results.map(extractTicketInfo).filter(Boolean);
```

Comment:
This would essentially also return multiple tickets if the response would yield multiple items in the `results` object. As `scan` is supposed to return a collection, I found this convenient. But it may also be a source for hidden problems in the future 🤔

Reply:
🤔 Indeed. My initial thought when writing the proposal was to `results.filter` and match by the UUID that we've constructed before mapping. Something along the lines of:

```js
function getTickets(response, id) {
  return (response.results || []).filter(r => r.id === id).map(extractTicketInfo);
}
```

Reply:
This means we're tightening the contract for `scan` to in the success case always return an array with only a single item. Sounds good and is probably a better safety-net for discovering changes to their API or general problems with our adapter.

Will update accordingly 🆗

Reply:
Implemented your suggestion _(thanks for the tip to skip the object deconstruction 🎉 )_ but kept the filter for null values.

Background: the ID is nested in the `value` object of the `result`. So filtering before extracting the ticket information would already do most of the work that's done when actually extracting the information. What do you think?

Reply:
Looks neat 👌